include/stl2/detail/range/access.hpp

```diff
@@ -474,76 +501,58 @@ STL2_OPEN_NAMESPACE {
 		template <class R>
 		constexpr bool has_member = false;
 		template <class R>
-		requires
-			requires(R& r) {
-				requires _Is<__f<decltype(r.data())>, std::is_pointer>;
+		requires requires(R& r) {
+				r.data();
+				requires std::is_pointer<__f<decltype(r.data())>>::value;
 			}
-		constexpr bool has_member<R> = true;
+		constexpr bool has_member<R&> = true;
 
 		template <class R>
 		constexpr bool has_pointer_iterator = false;
 		template <class R>
-		requires
-			requires(R& r) {
-				requires is_pointer<decltype(__stl2::begin(r))>::value;
-			}
+		requires std::is_pointer<__begin_t<R>>::value
 		constexpr bool has_pointer_iterator<R> = true;
 
 		template <class R>
 		constexpr bool has_contiguous_iterator = false;
 		template <class R>
-		requires
-			requires(R& r) {
-				requires ext::ContiguousIterator<decltype(__stl2::begin(r))>;
-			}
+		requires ext::ContiguousIterator<__begin_t<R>>
 		constexpr bool has_contiguous_iterator<R> = true;
 
 		struct fn {
 			// Prefer member
 			template <class R>
-			requires has_member<R>
+			requires has_member<R&>
 			constexpr auto operator()(R& r) const
 			STL2_NOEXCEPT_RETURN(
 				r.data()
 			)
-
 			// Return begin(r) if it's a pointer
 			template <class R>
 			requires !has_member<R> && has_pointer_iterator<R>
-			constexpr auto operator()(R& r) const
+			constexpr auto operator()(R&& r) const
 			STL2_NOEXCEPT_RETURN(
-				__stl2::begin(r)
+				__stl2::begin((R&&) r)
 			)
-
 			// Extension: Support contiguous ranges with non-pointer
 			//            iterators.
 			template <class R>
 			requires !has_member<R> && !has_pointer_iterator<R> &&
 				has_contiguous_iterator<R>
-			constexpr auto operator()(R& r) const
-			noexcept(noexcept(__stl2::begin(r) == __stl2::end(r)
-				? nullptr : detail::addressof(*__stl2::begin(r))))
+			constexpr auto operator()(R&& r) const
+			noexcept(noexcept(*declval<__begin_t<R>&>()) &&
+				noexcept(declval<__begin_t<R>&>() == declval<__end_t<R>>()))
```

Comment:
Not equivalent: this formulation critically doesn't include the calls to `begin` and `end`. (Once again, I really want `noexcept(auto)`.)